rfq/negotiator.go

```diff
@@ -0,0 +1,424 @@
+			err = fmt.Errorf("negotiator failed to add message "+
+				"to the outgoing messages channel (msg=%v)",
+				outgoingMsgResponse)
+			n.cfg.ErrChan <- err
+		}
+	}()
+
+	return nil
+}
+
+// SellOffer is a struct that represents an asset sell offer. This
+// data structure describes the maximum amount of an asset that is available
+// for sale.
+//
+// A sell offer is passive (unlike a buy order), meaning that it does not
+// actively lead to a bid request from a peer. Instead, it is used by the node
+// to selectively accept or reject incoming quote requests early before price
+// considerations.
+type SellOffer struct {
+	// AssetID represents the identifier of the subject asset.
+	AssetID *asset.ID
+
+	// AssetGroupKey is the public group key of the subject asset.
+	AssetGroupKey *btcec.PublicKey
+
+	// MaxUnits is the maximum amount of the asset under offer.
+	MaxUnits uint64
+}
+
+// Validate validates the asset sell offer.
+func (a *SellOffer) Validate() error {
+	if a.AssetID == nil && a.AssetGroupKey == nil {
+		return fmt.Errorf("asset ID is nil and asset group key is nil")
+	}
+
+	if a.AssetID != nil && a.AssetGroupKey != nil {
+		return fmt.Errorf("asset ID and asset group key are both set")
+	}
+
+	if a.MaxUnits == 0 {
+		return fmt.Errorf("max asset amount is zero")
+	}
+
+	return nil
+}
+
+// UpsertAssetSellOffer upserts an asset sell offer. If the offer already exists
+// for the given asset, it will be updated.
+func (n *Negotiator) UpsertAssetSellOffer(offer SellOffer) error {
+	// Validate the offer.
+	err := offer.Validate()
+	if err != nil {
+		return fmt.Errorf("invalid asset sell offer: %w", err)
+	}
+
+	n.sellOffersMtx.Lock()
+	defer n.sellOffersMtx.Unlock()
+
+	// Store the offer in the appropriate map.
+	//
+	// If the asset group key is not nil, then we will use it as the key for
+	// the offer. Otherwise, we will use the asset ID as the key.
+	switch {
+	case offer.AssetGroupKey != nil:
+		n.assetGroupSellOffers[*offer.AssetGroupKey] = offer
+
+	case offer.AssetID != nil:
+		n.assetSellOffers[*offer.AssetID] = offer
+	}
+
+	return nil
+}
+
+// RemoveAssetSellOffer removes an asset sell offer from the negotiator.
+func (n *Negotiator) RemoveAssetSellOffer(assetID *asset.ID,
+	assetGroupKey *btcec.PublicKey) error {
+
+	n.sellOffersMtx.Lock()
+	defer n.sellOffersMtx.Unlock()
+
+	// Remove the offer from the appropriate map.
+	//
+	// If the asset group key is not nil, then we will use it as the key for
+	// the offer. Otherwise, we will use the asset ID as the key.
+	switch {
+	case assetGroupKey != nil:
+		delete(n.assetGroupSellOffers, *assetGroupKey)
+
+	case assetID != nil:
+		delete(n.assetSellOffers, *assetID)
+
+	default:
+		return fmt.Errorf("asset ID and asset group key are both nil")
+	}
+
+	return nil
+}
+
+// HasAssetSellOffer returns true if the negotiator has an asset sell offer
+// which matches the given asset ID/group and asset amount.
+func (n *Negotiator) HasAssetSellOffer(assetID *asset.ID,
```

Comment:
nit: similar note from elsewhere re: distinguishing between non-existent sell offer vs. undersized sell offer.

Reply:
Good point, thanks!

I'm going to leave a TODO in this PR to modify `HasAssetSellOffer` (+ rename) such that it returns an error which will allow the caller to respond with a more informative reject message. And then I can have an error type for "undersized sell offer".